File: backend/api/sitio3/Nutricion.py
# backend/api/sitio3/Nutricion.py
from typing import List

from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session

from backend.database import get_db
from backend.models.sitio3.Nutricion import (
    NutricionS3Repository,
    NutricionS3Read,
    NutricionS3Create,
    NutricionS3Update,
)

router = APIRouter(
    prefix="/sitio3/nutricion",
    tags=["Sitio 3 - Nutrición"],
)


def get_repo(db: Session = Depends(get_db)) -> NutricionS3Repository:
    return NutricionS3Repository(db)


@router.get(
    "/",
    response_model=List[NutricionS3Read],
    summary="Listar registros nutricionales Sitio 3",
)
def listar_nutricion_s3(
    empresa_id: int,
    granja_id: int,
    repo: NutricionS3Repository = Depends(get_repo),
):
    return repo.listar_por_granja(empresa_id, granja_id)


@router.post(
    "/",
    response_model=NutricionS3Read,
    status_code=status.HTTP_201_CREATED,
    summary="Registrar nutrición Sitio 3",
)
def crear_nutricion_s3(
    payload: NutricionS3Create,
    repo: NutricionS3Repository = Depends(get_repo),
):
    if payload.alimento_consumido < 0 or payload.cantidad_suplemento < 0:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Alimento y suplemento no pueden ser negativos",
        )
    return repo.crear(payload)


@router.put(
    "/{reg_id}",
    response_model=NutricionS3Read,
    summary="Actualizar nutrición Sitio 3",
)
def actualizar_nutricion_s3(
    reg_id: int,
    empresa_id: int,
    granja_id: int,
    cambios: NutricionS3Update,
    repo: NutricionS3Repository = Depends(get_repo),
):
    reg = repo.obtener_por_id(reg_id, empresa_id, granja_id)
    if not reg:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Registro nutricional Sitio 3 no encontrado",
        )

    data = cambios.dict(exclude_unset=True)
    nuevo_alim = data.get("alimento_consumido", reg.alimento_consumido)
    nuevo_supl = data.get("cantidad_suplemento", reg.cantidad_suplemento)
    if nuevo_alim < 0 or nuevo_supl < 0:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Alimento y suplemento no pueden ser negativos",
        )

    return repo.actualizar(reg, cambios)


@router.delete(
    "/{reg_id}",
    status_code=status.HTTP_204_NO_CONTENT,
    summary="Eliminar registro nutricional Sitio 3",
)
def eliminar_nutricion_s3(
    reg_id: int,
    empresa_id: int,
    granja_id: int,
    repo: NutricionS3Repository = Depends(get_repo),
):
    reg = repo.obtener_por_id(reg_id, empresa_id, granja_id)
    if not reg:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Registro nutricional Sitio 3 no encontrado",
        )
    repo.eliminar(reg)

File: frontend/src/services/sitio3/Nutricion.ts
import API_BASE from "../../config/api";

export type RegistroNutricion3 = {
  id: number;
  fecha: string;
  corral: string;
  lote: string;
  dieta: string;
  alimento_consumido: number;
  suplemento: string;
  cantidad_suplemento: number;
  responsable: string;
  observaciones: string;
};

export type NuevoRegistroNutricion3 = Omit<RegistroNutricion3, "id">;

const EMPRESA_ID = 1;
const GRANJA_ID = 1;

function mapApiToNut3(api: any): RegistroNutricion3 {
  return {
    id: api.id,
    fecha: api.fecha,
    corral: api.corral,
    lote: api.lote,
    dieta: api.dieta,
    alimento_consumido: api.alimento_consumido,
    suplemento: api.suplemento,
    cantidad_suplemento: api.cantidad_suplemento,
    responsable: api.responsable,
    observaciones: api.observaciones ?? "",
  };
}

export async function getNutricionSitio3(): Promise<RegistroNutricion3[]> {
  const res = await fetch(
    `${API_BASE}/sitio3/nutricion/?empresa_id=${EMPRESA_ID}&granja_id=${GRANJA_ID}`
  );
  if (!res.ok) {
    throw new Error("Error al obtener registros nutricionales Sitio 3");
  }
  const data = await res.json();
  return data.map(mapApiToNut3);
}

export async function addNutricionSitio3(
  reg: NuevoRegistroNutricion3
): Promise<RegistroNutricion3> {
  const payload = {
    empresa_id: EMPRESA_ID,
    granja_id: GRANJA_ID,
    ...reg,
  };

  const res = await fetch(`${API_BASE}/sitio3/nutricion/`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(payload),
  });
  if (!res.ok) {
    throw new Error("Error al registrar nutrición Sitio 3");
  }
  const data = await res.json();
  return mapApiToNut3(data);
}

export async function updateNutricionSitio3(
  id: number,
  reg: NuevoRegistroNutricion3
): Promise<RegistroNutricion3> {
  const res = await fetch(
    `${API_BASE}/sitio3/nutricion/${id}?empresa_id=${EMPRESA_ID}&granja_id=${GRANJA_ID}`,
    {
      method: "PUT",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(reg),
    }
  );
  if (!res.ok) {
    throw new Error("Error al actualizar nutrición Sitio 3");
  }
  const data = await res.json();
  return mapApiToNut3(data);
}

export async function deleteNutricionSitio3(id: number): Promise<void> {
  const res = await fetch(
    `${API_BASE}/sitio3/nutricion/${id}?empresa_id=${EMPRESA_ID}&granja_id=${GRANJA_ID}`,
    { method: "DELETE" }
  );
  if (!res.ok) {
    throw new Error("Error al eliminar registro nutricional Sitio 3");
  }
}
